apps/sms/test/marionette/lib/storage.js

```diff
@@ -57,15 +27,29 @@
          * pre-populate storage with.
          */
         setContactsStorage: function(contacts) {
-          client.executeScript(function(contacts) {
-            var STORAGE_NAME = 'contactsDB';
-
-            window.wrappedJSObject.TestStorages.setStorage(STORAGE_NAME, {
-              contacts: contacts
-            });
+          this.setStorage('contactsDB', {
+            contacts: contacts || []
+          });
+        },
 
-            window.wrappedJSObject.TestStorages.setStorageReady(STORAGE_NAME);
-          }, [contacts || []]);
+        /**
+         * Sets pre-populated custom storage.
+         * @param {string} storageName Name of the custom storage.
+         * @param {*} storage Custom storage data.
+         * @private
+         */
+        setStorage: function(storageName, storage) {
+          client.scope({ context: 'chrome' }).executeScript(
+            function(storage, storageName) {
+              Services.obs.addObserver(function(window) {
+                window.wrappedJSObject.TestStorages.set(
+                  storageName,
+                  Components.utils.cloneInto(storage, window.wrappedJSObject)
+                );
+              }, 'test-storage-ready', false);
```

Comment:
> it's not clear to me how many times the test-storage-ready can be sent here :)

I'm planning to improve our mocks further in the future - currently I see our mocks are registered several times - it's ok that they are registered for keyboard and Homescreen, but it's not obvious to me why our mocks are registered twice for main frame and twice for shim host - same will be for this event - it will fire several times and every time storage will be rewritten - it's old problem that I didn't find time to fix yet :)